# Projeto 1 - Ciência dos Dados

Nome: Breno Alencar Araújo

Nome: Celina Vieira de Melo

Atenção: Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [542]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

In [543]:
print('Esperamos trabalhar no diretório')
print(os.getcwd())

Esperamos trabalhar no diretório
C:\Users\Breno Alencar\Documents\Insper\Semestre 2\CDados\proj2copia2


Carregando a base de dados com os tweets classificados como relevantes e não relevantes:

In [544]:
filename = 'Windows 10_final.xlsx'

No projeto 1 escolhemos a marca Windows 10 para fazer a análise dos Tweets para a empresa! Com isso, classificamos como relevante os Tweets que estavam relacionados as opiniões dos usuários em relação ao produto e como irrelevantes anúncios, links para correção de erros do Windows etc.

In [545]:
#Criando Dataframe da planilha Treinamento
train = pd.read_excel(filename)
train.head(5)

Treinamento  Relevância
0  relógio casio que eu não usava há meses com a ...           1
1  rt @pplware: hoje trazemos 3 ferramentas que o...           0
2  ct news - 15/03/2021 (windows 10 começa a dar ...           1
3  quer comprar este produto?entao clique neste l...           0
4  você ainda pode fazer upgrade gratuito para o ...           1

In [546]:
#Criando Dataframe da aba Teste da planilha
test = pd.read_excel(filename, sheet_name = 'Teste')
test.head(5)

Teste  Relevância
0   um dia eu restauro um windows 10 outro dia um xp           1
1  rt @ajulianafiuza: vivendo ou atualizando o wi...           1
2  finalmente voltei o meu pc p windows 10 obg se...           1
3     @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1
4  todo dia esse windows 10 pedindo atualização, ...           1

In [587]:
#Pd Series de Treinamento
treinamento = pd.Series(train.Treinamento)


0      relógio casio que eu não usava há meses com a ...
1      rt @pplware: hoje trazemos 3 ferramentas que o...
2      ct news - 15/03/2021 (windows 10 começa a dar ...
3      quer comprar este produto?entao clique neste l...
4      você ainda pode fazer upgrade gratuito para o ...
                             ...                        
295    @microsoftbr @microsoftajuda não consigo ajuda...
296    você pode ter o pc com os melhores specs do mu...
297    @clapiclapiclapi esse daqui é um i5, 8gb de ra...
298    windows 10 recebe correção temporária da micro...
299    rt @magazinewmti: notebook hp 250 g7 intel cor...
Name: Treinamento, Length: 300, dtype: object

___
## Classificador automático de sentimento


O Windows 10 é o mais novo sistema operacional da Microsft e desde seu lançamento está rodeado de opiniões díspares.  
Para nosso classificador Naive-Bayes, consideramos como relevantes as constantes atualizações que o sistema passa, o tempo da atualização e o desempenho em jogos, além das experiências dos usuários do sistema. 

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [548]:
import re 

def limpeza(text):
    """
        Função de limpeza que troca alguns sinais básicos por espaços
    """
    #importando string
    punctuation = '[!-.:?;''""@"|"\n"/]' # Sinais [] como delimitadores de um conjunto.
    pattern = re.compile(punctuation)
    text_subbed = re.sub(pattern, '', text)
    return text_subbed

In [549]:
#Removendo espaços dos Tweets Relevantes e Irrelevantes

classifi_0 =' '.join(train.loc[filt_0,'Treinamento'])

#A função .lower() deixa todos os caracteres em minúsculo pois caracteres
# maiússculos não acrescentam em nada na nossa análise, além de o Python contar duas vezes a mesma palavra. 

classifi_0_limpo = limpeza(classifi_0.lower())

classifi_1_limpo = limpeza(classifi_1.lower())


In [550]:
#Filtrando os Tweets Relevantes e Irrelevantes da aba Treinamento

filt_1 = train.Relevância == 1
classifi_1 =' '.join(train.loc[filt_1,'Treinamento'])


filt_0 = train.Relevância == 0
classifi_0 =' '.join(train.loc[filt_0,'Treinamento'])



In [551]:
#Todas as palavras separadas dos tweets como Relevantes e Irrelevantes

todas_os_tweets_irrelevantes = classifi_0_limpo.split()

todas_os_tweets_relevantes = classifi_1_limpo.split()

In [552]:
#PARA OS TWEETS RELEVANTES

#Criando um pd.serie para todos os Tweets Relevantes
serie_tweets_r = pd.Series(todas_os_tweets_relevantes)

#Criando um pd.serie para todos os Tweets Relevantes
serie_tweets_r = pd.Series(todas_os_tweets_relevantes)

#Criando Tabela de Frequência absoluta dos Tweets Relevantes
tabela_tweets_r = serie_tweets_r.value_counts()

In [553]:
#PARA OS TWEETS IRRELEVANTES

#Criando um pd.serie para todos os Tweets Irrelevantes
serie_tweets_irr = pd.Series(todas_os_tweets_irrelevantes)

#Criando um pd.serie para todos os Tweets Irrelevantes
serie_tweets_irr = pd.Series(todas_os_tweets_irrelevantes)

#Criando Tabela de Frequência absoluta dos Tweets Irrelevantes
tabela_tweets_irr = serie_tweets_irr.value_counts()


In [554]:
#Criando uma tabela relativa para os Twwts Relevantes 

tabela_tweets_relativa_r = serie_tweets_r.value_counts(True)

#Criando uma tabela relativa para os Twwts Irrelevantes 

tabela_tweets_relativa_irr = serie_tweets_irr.value_counts(True)


In [577]:
#Criando uma nova variável que contem todas as palavras dos Tweets Relevantes e Irrelevantes da planilha treinamento 
palavras_tweets = classifi_0_limpo + classifi_1_limpo


In [579]:
#Colocando todas as palavras dos Tweets em uma lista

todas_os_tweets = palavras_tweets.split()

#Transformando em pd.Series para podermos trabalharmos com as ferramentas do Pandas 

serie_tweets = pd.Series(todas_os_tweets)

In [581]:
# tabela_relativo_tweets = serie_tweets.value_counts(True)
tabela_absoluta_tweets = serie_tweets.value_counts()
tabela_absoluta_tweets.head(10)

windows    291
10         256
o          181
de         179
que        128
e          125
a          116
do         111
é           82
no          80
dtype: int64

In [582]:
#Descobrindo a quantidade de palavras sem repetição de todos os Tweets

tabela_absoluta_tweets.shape[0]

2111

In [583]:
#Calculando as probabilidades de um Tweet ser relevante dentre todos os Tweets 

probabilidade_relevante = len(todas_os_tweets_relevantes)/len(todas_os_tweets)
probabilidade_relevante

0.6376153966515412

In [584]:
#Calculando as probabilidades de um Tweet ser irrelevante dentre todos os Tweets 

probilidade_irrelevantes = len(todas_os_tweets_irrelevantes)/len(todas_os_tweets)
probilidade_irrelevantes

0.36254107338444685

Com base nas probabilidades calculadas, conclui-se que a probabilidade do Tweet ter relevante é maior do que ser irrelevante mas ainda sim temos uma quantidade quase equilibrada entre a quantidade total de Tweets.

___
### Verificando a performance do Classificador

Agora você deve testar o seu classificador com a base de Testes.

In [565]:
#Tamanho de Pd Series de Teste

quantidade_de_tweets = len(test['Teste'])

In [586]:
#Pd Series de Teste

frase = (test['Teste'])

0       um dia eu restauro um windows 10 outro dia um xp
1      rt @ajulianafiuza: vivendo ou atualizando o wi...
2      finalmente voltei o meu pc p windows 10 obg se...
3         @doliveira__ esse aqui https://t.co/fpcbvgkkhj
4      todo dia esse windows 10 pedindo atualização, ...
                             ...                        
203    olha, ter um computador com windows 10 sem ssd...
204    todo o meu trabalho naquele ssd e windows 10 i...
205    anotar pra não esquecer:\nnunca mais tentar in...
206    @guilhox pois é, o windows 10 mesmo eu uso obr...
207    você pode ter o pc com os melhores specs do mu...
Name: Teste, Length: 208, dtype: object


In [589]:
#Transformando Pd Series de Teste em lista
lista = list(frase)

lista_1 = []

j = 0
while j < len(lista):
    lista_1.append(lista[j])
    j+=1

#Trocando frase Serie para lista criada
frase = lista_1


In [590]:
#Procedimento para suaviazaação de Laplace (quando necessário), sendo esse acrescentado às novas listas os resultados obtidos.

lista_relevante = []
lista_irrelevante = []

for i in frase:
    i = i.split()
    for palavra in i:
        
        probFraseDadoR = 1
        #Caso uma palavra esteja na tabela de tweets relevantes
        if palavra in tabela_tweets_r:
            probFraseDadoR *= (tabela_tweets_r[palavra] + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_relevantes) )
            
        else:
            #Caso contrário, tomaremos a suavização de Laplace para não zerarmos um dos termos 
            probFraseDadoR *= (0 + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_relevantes) )
            
                
    lista_relevante.append(probFraseDadoR) 
#Fizemos o análogo para os irrelevantes
for k in frase:
    k = k.split()
    for palavra in k:
        
        probFraseDadoIRR = 1
        
        if palavra in tabela_tweets_irr:
            probFraseDadoIRR *= (tabela_tweets_irr[palavra] + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_irrelevantes) )
        
        else: 
            probFraseDadoIRR *= (0 + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_irrelevantes) )
         
        
    lista_irrelevante.append(probFraseDadoIRR)

In [591]:
test.head(5)

Teste  Relevância  \
0   um dia eu restauro um windows 10 outro dia um xp           1   
1  rt @ajulianafiuza: vivendo ou atualizando o wi...           1   
2  finalmente voltei o meu pc p windows 10 obg se...           1   
3     @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1   
4  todo dia esse windows 10 pedindo atualização, ...           1   

   Classificador  
0              1  
1              1  
2              1  
3              1  
4              1

In [572]:

probabilidade_relevante = len(todas_os_tweets_relevantes)/len(todas_os_tweets)
probabilidade_irrelevante = len(todas_os_tweets_irrelevantes)/len(todas_os_tweets)
lista_classificador = []


contador = 0
while contador < len(lista_relevante):
    
    Prob_de_ser_Relevante = probabilidade_relevante*lista_relevante[contador]
    Prob_de_ser_Irrelevante = probabilidade_irrelevante*lista_irrelevante[contador]
  
 #Para finalizarmos, fazemos a comparação final das probabilidades, classificamos como relevante ou irrelevante e adicionamos 
 #essa classificação em uma nova lista.

    if Prob_de_ser_Relevante > Prob_de_ser_Irrelevante:
        lista_classificador.append(1)
    
    else:
        lista_classificador.append(0)
        
    contador +=1
        

test["Classificador"] = lista_classificador
test.head()

Teste  Relevância  \
0   um dia eu restauro um windows 10 outro dia um xp           1   
1  rt @ajulianafiuza: vivendo ou atualizando o wi...           1   
2  finalmente voltei o meu pc p windows 10 obg se...           1   
3     @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1   
4  todo dia esse windows 10 pedindo atualização, ...           1   

   Classificador  
0              1  
1              1  
2              1  
3              1  
4              1

In [573]:
#Criando tabela para comparação entre dados classificados e saída do calssificador

pd.crosstab(test["Relevância"],test["Classificador"],normalize=True)*100

Classificador         0          1
Relevância                        
0              3.846154  32.692308
1              0.000000  63.461538

In [592]:
tabela_tweets_relativa_r.head()

windows    0.048834
10         0.043436
o          0.032393
de         0.026503
que        0.023313
dtype: float64

___
### Concluindo

   A partir da tabela CrosTab concluímos um acerto de 63,46% dos casos em que marcamos como relevante e um acerto de 3,36% dos casos em que marcamos como irrelevante

- Verdadeiro positivo: 63,46%

- Falso positivo: 32.69%

- Falso negativo: 0.0%

- Verdadeiro negativo: 3.84%

Mais de um terço dos tweets foi classificado como falso positivo devido a fatores que são difíceis para o classificador entender como ironia e ambiguidade. Vide exemplo: "nem a existência do windows 10 foi uma ideia tão idiota assim ". 


Vale ressaltar o porque não podemos alimentar nosso base de Treinamento automaticamente usando o próprio classificador, aplicado a novos tweets. Isso se deve as classificações que fizemos! Pode-se dizer que estaríamos misturando resultados prováveis a classificacões corretas. Com o passar das iterações, as probabilidades irão reduzir cada vez mais, tornando mais inseguro confiar na classificação feita pelo classificador. 

Outros cenários possíveis são aplicações na medicina, no diagnóstico de doenças a partir dos sintomas relatados pelos pacientes, filtragem de SPAMs também é possível com o Naive Bayes ou a classificação de textos com base em um banco de dados. https://www.datageeks.com.br/naive-bayes/#:~:text=Hoje%20%C3%A9%20tamb%C3%A9m%20utilizado%20na,com%20base%20em%20seu%20conte%C3%BAdo.

Para melhorar o classificador poderíamos pegar uma base de dados maior, com mais tweets, remover preposições, pois não trazem informações relevantes para classificação e poderíamos implementar uma classificação intermediária, por exemplo. Para isso, fariamos uma nova filtragem e calcularíamos as novas probabilidades para ter resultados mais precisos.
https://repositorio.ufmg.br/bitstream/1843/ALDR-73WQXZ/1/tese1.pdf (Ctrl+f buscar por 'agrega')

___
### Qualidade do Classificador a partir de novas separações dos tweets entre Treinamento e Teste

Caso for fazer esse item do Projeto

In [593]:
#A partir daqui mesclaremos os tweets de treinamento e teste, para nova combinação de probabilidades e análise de um novo cenário.
precisao_percent = []
train2 = pd.read_excel(filename)
test2 = pd.read_excel(filename, sheet_name='Teste')
train2 = train2.rename(columns={'Treinamento':'Todos'})
test2 = test2.rename(columns={'Teste':'Todos'})
total = pd.concat([test2,train2])
total.head()

Todos  Relevância
0     um dia eu restauro um windows 10 outro dia um xp           1
1    rt @ajulianafiuza: vivendo ou atualizando o wi...           1
2    finalmente voltei o meu pc p windows 10 obg se...           1
3       @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1
4    todo dia esse windows 10 pedindo atualização, ...           1
..                                                 ...         ...
295  @microsoftbr @microsoftajuda não consigo ajuda...           1
296  você pode ter o pc com os melhores specs do mu...           1
297  @clapiclapiclapi esse daqui é um i5, 8gb de ra...           0
298  windows 10 recebe correção temporária da micro...           1
299  rt @magazinewmti: notebook hp 250 g7 intel cor...           0

[508 rows x 2 columns]

___
## Aperfeiçoamento:

Os trabalhos vão evoluir em conceito dependendo da quantidade de itens avançados:

* Limpar: \n, :, ", ', (, ), etc SEM remover emojis
* Corrigir separação de espaços entre palavras e emojis ou entre emojis e emojis
* Propor outras limpezas e transformações que não afetem a qualidade da informação ou classificação
* Criar categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante (3 categorias: C, mais categorias conta para B)
* Explicar por que não posso usar o próprio classificador para gerar mais amostras de treinamento
* Propor diferentes cenários para Naïve Bayes fora do contexto do projeto
* Sugerir e explicar melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)


___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**